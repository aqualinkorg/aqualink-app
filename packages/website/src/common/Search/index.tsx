import React, { useState, ChangeEvent } from "react";
import { useDispatch, useSelector } from "react-redux";
import {
  withStyles,
  WithStyles,
  createStyles,
  Grid,
  IconButton,
  TextField,
} from "@material-ui/core";
import SearchIcon from "@material-ui/icons/Search";
import Autocomplete from "@material-ui/lab/Autocomplete";

import { setReefOnMap } from "../../store/Homepage/homepageSlice";
import type { Reef } from "../../store/Reefs/types";
import { reefsListSelector } from "../../store/Reefs/reefsListSlice";

const Search = ({ classes }: SearchProps) => {
  const [searchedReef, setSearchedReef] = useState<Reef | null>(null);
  const dispatch = useDispatch();
  const reefs = useSelector(reefsListSelector);

  const onChangeSearchText = (
    event: ChangeEvent<HTMLTextAreaElement | HTMLInputElement>
  ) => {
    const searchValue = event.target.value;
    const index = reefs.findIndex(
      (reef) => reef.name?.toLowerCase() === searchValue.toLowerCase()
    );
    if (index > -1) {
      setSearchedReef(reefs[index]);
    }
  };

  const onDropdownItemSelect = (event: ChangeEvent<{}>, value: Reef | null) => {
    if (value) {
      setSearchedReef(value);
      dispatch(setReefOnMap(value));
    }
  };

  const onSearchSubmit = () => {
    if (searchedReef) {
      dispatch(setReefOnMap(searchedReef));
    }
  };

  return (
    <Grid container justify="flex-end" item xs={6}>
      <Grid
        className={classes.searchBar}
        container
        alignItems="center"
        item
        xs={8}
      >
        <Grid
          className={classes.searchBarIcon}
          item
          xs={2}
          container
          alignItems="center"
          justify="center"
        >
          <IconButton size="small" onClick={onSearchSubmit}>
            <SearchIcon />
          </IconButton>
        </Grid>
        <Grid
          className={classes.searchBarText}
          item
          xs={10}
          container
          alignItems="center"
        >
          <Autocomplete
            id="location"
            className={classes.searchBarInput}
            options={reefs}
            getOptionLabel={(reef) => reef.name || ""}
            value={searchedReef}
            onChange={onDropdownItemSelect}
            onInputChange={(event, value, reason) =>
              reason === "clear" && setSearchedReef(null)
            }
            renderInput={(params) => (
              <TextField
                {...params}
                onChange={onChangeSearchText}
                style={{ height: "100%" }}
                placeholder="Search Location"
                variant="outlined"
                InputLabelProps={{
                  shrink: false,
                }}
              />
            )}
          />
        </Grid>
      </Grid>
    </Grid>
  );
};

const styles = () =>
  createStyles({
    searchBar: {
      height: 42,
    },
    searchBarIcon: {
      backgroundColor: "#6ba8c0",
      borderRadius: "4px 0 0 4px",
      height: "100%",
    },
    searchBarText: {
      paddingLeft: "0.5rem",
      backgroundColor: "#469abb",
      borderRadius: "0 4px 4px 0",
      height: "100%",
    },
    searchBarInput: {
      "& .MuiOutlinedInput-root .MuiOutlinedInput-notchedOutline": {
        borderWidth: 0,
      },
      "&:hover .MuiOutlinedInput-root .MuiOutlinedInput-notchedOutline": {
        borderWidth: 0,
      },
      "& .MuiOutlinedInput-root.Mui-focused .MuiOutlinedInput-notchedOutline": {
        borderWidth: 0,
      },
      height: "100%",
      width: "100%",
    },
  });

type SearchProps = WithStyles<typeof styles>;

export default withStyles(styles)(Search);
