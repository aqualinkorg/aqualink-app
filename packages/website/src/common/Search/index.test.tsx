import React from "react";
import { Provider } from "react-redux";
import { render } from "@testing-library/react";
import configureStore from "redux-mock-store";

import Search from ".";

const mockStore = configureStore([]);

describe("Search", () => {
  let element: HTMLElement;
  beforeEach(() => {
    const store = mockStore({
      reefsList: {
        list: [],
        loading: false,
        error: null,
      },
    });

    store.dispatch = jest.fn();

    element = render(
      <Provider store={store}>
        <Search />
      </Provider>
    ).container;
  });

  it("should render with given state from Redux store", () => {
    expect(element).toMatchSnapshot();
  });
});
