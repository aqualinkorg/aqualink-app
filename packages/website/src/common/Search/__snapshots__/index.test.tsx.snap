// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search should render with given state from Redux store 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-justify-xs-flex-end MuiGrid-grid-xs-6"
  >
    <div
      class="MuiGrid-root Search-searchBar-1 MuiGrid-container MuiGrid-item MuiGrid-align-items-xs-center MuiGrid-grid-xs-8"
    >
      <div
        class="MuiGrid-root Search-searchBarIcon-2 MuiGrid-container MuiGrid-item MuiGrid-align-items-xs-center MuiGrid-justify-xs-center MuiGrid-grid-xs-2"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiGrid-root Search-searchBarText-3 MuiGrid-container MuiGrid-item MuiGrid-align-items-xs-center MuiGrid-grid-xs-10"
      >
        <div
          aria-expanded="false"
          class="MuiAutocomplete-root Search-searchBarInput-4 MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon"
          role="combobox"
        >
          <div
            class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
            style="height: 100%;"
          >
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiAutocomplete-inputRoot MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
            >
              <input
                aria-autocomplete="list"
                aria-invalid="false"
                autocapitalize="none"
                autocomplete="off"
                class="MuiInputBase-input MuiOutlinedInput-input MuiAutocomplete-input MuiAutocomplete-inputFocused MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
                id="location"
                placeholder="Search Location"
                spellcheck="false"
                type="text"
                value=""
              />
              <div
                class="MuiAutocomplete-endAdornment"
              >
                <button
                  aria-label="Clear"
                  class="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-clearIndicator"
                  tabindex="-1"
                  title="Clear"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
                <button
                  aria-label="Open"
                  class="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-popupIndicator"
                  tabindex="-1"
                  title="Open"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M7 10l5 5 5-5z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
              </div>
              <fieldset
                aria-hidden="true"
                class="PrivateNotchedOutline-root-5 MuiOutlinedInput-notchedOutline"
                style="padding-left: 8px;"
              >
                <legend
                  class="PrivateNotchedOutline-legend-6"
                  style="width: 0.01px;"
                >
                  <span>
                    â€‹
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
